pkg/collector/container.go

```diff
@@ -15,24 +15,56 @@
 package collector
 
 import (
+	"errors"
 	"fmt"
+	"net"
+	"sort"
+	"strconv"
 
 	"github.com/go-logr/logr"
+	"github.com/mitchellh/mapstructure"
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/validation"
 
 	"github.com/open-telemetry/opentelemetry-operator/apis/v1alpha1"
 	"github.com/open-telemetry/opentelemetry-operator/internal/config"
 	"github.com/open-telemetry/opentelemetry-operator/pkg/collector/adapters"
 	"github.com/open-telemetry/opentelemetry-operator/pkg/naming"
 )
 
+const maxPortLen = 15
+
+var errInvalidPort = errors.New("invalid port name/num")
+
 // Container builds a container for the given collector.
 func Container(cfg config.Config, logger logr.Logger, otelcol v1alpha1.OpenTelemetryCollector) corev1.Container {
 	image := otelcol.Spec.Image
 	if len(image) == 0 {
 		image = cfg.CollectorImage()
 	}
 
+	// build container ports from service ports
+	ports := getConfigContainerPorts(logger, otelcol.Spec.Config)
+	for _, p := range otelcol.Spec.Ports {
+		truncName := naming.Truncate(p.Name, maxPortLen)
+		if p.Name != truncName {
+			logger.Info("truncating container port name",
+				"port.name.prev", p.Name, "port.name.new", truncName)
+		}
+		nameErrs := validation.IsValidPortName(truncName)
+		numErrs := validation.IsValidPortNum(int(p.Port))
+		if len(nameErrs) > 0 || len(numErrs) > 0 {
+			logger.Error(errInvalidPort, "dropping container port", "port.name", truncName, "port.num", p.Port,
```

Comment:
I wonder if we should rather emit k8s event to let the user know that the validation failed.

Reply:
@kevinearls you did some logging-related work recently. 

Do you think we should use k8s events here or normal logs?

Reply:
@pavolloffay I don't really know, to be honest.  I think logging would be fine but we are not very consistent about this.